dali/kernels/imgproc/warp/mapping_traits.h

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_WARP_MAPPING_TRAITS_H_
+#define DALI_KERNELS_IMGPROC_WARP_MAPPING_TRAITS_H_
+
+#include <functional>
+#include "dali/core/geom/vec.h"
+
+namespace dali {
+namespace kernels {
+namespace warp {
+
+std::true_type is_fp_mapping_helper(const std::function<vec2(vec2)> &f);
+std::true_type is_fp_mapping_helper(const std::function<vec3(vec3)> &f);
+std::true_type is_fp_mapping_helper(const std::function<vec4(vec4)> &f);
+
+std::false_type is_fp_mapping_helper(const std::function<ivec2(ivec2)> &f);
+std::false_type is_fp_mapping_helper(const std::function<ivec3(ivec3)> &f);
+std::false_type is_fp_mapping_helper(const std::function<ivec4(ivec4)> &f);
+
+
+template <typename Mapping>
+struct is_fp_mapping : decltype(is_fp_mapping_helper(Mapping())) {};
```

Comment:
Is this an inheritance from the type of the function? How does it actually work?

Reply:
This type is inheriting from a return type of a function - and that function is automatically selected by overload resolution. It's quite common in type traits.
How it works is:
The `Mapping` type must be usable as a function taking a vector and returning a vector. If it takes a `float` vector (`vecN`), then we consider the mapping to be "fp" (floating point). Otherwise, it must operate on `ivec` (not fp). If neither overload can be used, the mapping is ill-formed and the code should not compile.

Reply:
Do you actually need that std::function, or just overload on function pointer would be enough?

Reply:
`std::function` is required, because I want the mapping object to decay to something callable with given signature.